plugins/settings/settings_plugin.c

```diff
@@ -180,53 +266,209 @@ settings_plugin_get_boolean(
+}
+
+static
+gboolean
+settings_plugin_update_settings(
+    SettingsPlugin* self,
+    GKeyFile* config)
+{
+    GStrV* plugins = self->order;
     gboolean save = FALSE;
 
-    if (enabled != self->nfc_enabled) {
+    if (settings_plugin_update_boolean(self, config, SETTINGS_KEY_ENABLED,
+        self->nfc_enabled)) {
         save = TRUE;
-        g_key_file_set_boolean(config, SETTINGS_GROUP, SETTINGS_KEY_ENABLED,
-            self->nfc_enabled);
     }
 
-    if (save) {
+    /* Check plugin configs */
+    if (plugins) {
+        while (*plugins) {
+            const char* group = *plugins++;
+            const SettingsPluginConfig* pc = g_hash_table_lookup(self->plugins,
+                group);
+            const char* const* keys = nfc_config_get_keys(pc->config);
+
+            if (keys) {
+                const char* const* ptr = keys;
+
+                while (*ptr) {
+                    const char* key = *ptr++;
+                    GVariant* value = nfc_config_get_value(pc->config, key);
+                    char* str = g_key_file_get_string(config, group, key, NULL);
+                    char* defval = g_key_file_get_string(self->defaults,
+                        group, key, NULL);
+                    char* sval = NULL;
+
+                    if (value) {
+                        sval = g_variant_print(value, FALSE);
+                        g_variant_unref(value);
+                    }
+
+                    if (sval && defval && !strcmp(sval, defval)) {
+                        /* Don't store the default value */
+                        GVERBOSE_("[%s] %s %s => (default)", group,
+                            key, sval);
+                        if (g_key_file_remove_key(config, group, key, NULL)) {
+                            save = TRUE;
+                        }
+                    } else if (g_strcmp0(sval, str)) {
+                        GVERBOSE_("[%s] %s %s => %s", group, key, str, sval);
+                        if (sval) {
+                            g_key_file_set_string(config, group, key, sval);
+                            save = TRUE;
+                        } else if (g_key_file_remove_key(config, group, key,
+                            NULL)) {
+                            save = TRUE;
+                        }
+                    }
+
+                    g_free(defval);
+                    g_free(sval);
+                    g_free(str);
+                }
+            }
+        }
+    }
+
+    return save;
+}
+
+static
+void
+settings_plugin_update_config(
+    SettingsPlugin* self)
+{
+    GKeyFile* config = settings_plugin_load_config(self);
+
+    if (settings_plugin_update_settings(self, config)) {
         settings_plugin_save_config(self, config);
     }
 
     g_key_file_unref(config);
 }
 
-#ifdef HAVE_DBUSACCESS
+static
+gboolean
+settings_plugin_is_valid_key(
+    NfcConfigurable* config,
+    const char* key)
+{
+    return gutil_strv_contains((char**)nfc_config_get_keys(config), key);
```

Comment:
Random observation: there is mixed use of GStrv and char** types

Reply:
(even more random observation, GStrV* (which is not the same as GStrv) and char**)

Reply:
Made the cast match `gutil_strv_contains()` declaration.